spec/02-integration/04-admin_api/06-certificates_routes_spec.lua

```diff
@@ -156,31 +176,310 @@ describe("Admin API: #" .. kong_config.database, function()
+        assert.equal(2, #json.data)
+        assert.equal(2, json.total)
+
+        -- make sure we did not add any certificate
+        res = assert(client:send {
+          method = "GET",
+          path = "/certificates",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, json.total)
+        assert.equal(2, #json.data)
+      end)
+
+      it("updates snis associated with a certificate", function()
+
+        local res = assert(client:send {
+          method  = "PUT",
+          path    = "/certificates",
+          body    = {
+            id = cert_foo.id,
+            snis  = "baz.com",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+
+        local body = assert.res_status(200, res)
+        local json = cjson.decode(body)
+
+        assert.same({"baz.com" }, json.snis)
+
+        -- make sure number of snis don't change
+        -- since we delete foo.com and added baz.com
+        res = assert(client:send {
+          method  = "GET",
+          path    = "/snis",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, #json.data)
+        assert.equal(2, json.total)
+
+        -- make sure we did not add any certificate
+        res = assert(client:send {
+          method = "GET",
+          path = "/certificates",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, json.total)
+        assert.equal(2, #json.data)
+      end)
+
+      it("updates only the certificate if no snis are specified", function()
+
+        local res = assert(client:send {
+          method  = "PUT",
+          path    = "/certificates",
+          body    = {
+            id = cert_bar.id,
+            cert  = "bar_cert",
+            key   = "bar_key",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+
+        local body = assert.res_status(200, res)
+        local json = cjson.decode(body)
+
+        -- make sure certificate got updated and sni remains the same
+        assert.same({"bar.com" }, json.snis)
+        assert.same("bar_cert", json.cert)
+        assert.same("bar_key", json.key)
+
+        -- make sure number of snis don't change
+        res = assert(client:send {
+          method  = "GET",
+          path    = "/snis",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, #json.data)
+        assert.equal(2, json.total)
+
+        -- make sure we did not add any certificate
+        res = assert(client:send {
+          method = "GET",
+          path = "/certificates",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, json.total)
+        assert.equal(2, #json.data)
```

Comment:
Could be worth checking if our updated certificate indeed has `bar_cert` and `bar_key` here instead of its previous values?

Reply:
We do check that in the response of the `PUT` on Line 355.
Are you saying that we should do another `GET` for `cert_bar.id` and make sure that the certificate was updated?

Reply:
> Are you saying that we should do another GET for cert_bar.id and make sure that the certificate was updated?

Yes, this is what I am saying. As of today, it so happens that our DAO will run subsequent queries for Cassandra to select the inserted value, or will make use of the `RETURNING *` feature of PostgreSQL. However, we cannot guarantee that this will be the case in the future (we are currently implementing a new DAO internallt that tries to avoid such read-before-write and read-after-write patterns already), nor for other databases that we may implement.